tools/s3bulk/main.go

```diff
@@ -0,0 +1,237 @@
+package main
+
+import (
+	"bytes"
+	"flag"
+	"fmt"
+	"os/exec"
+
+	"github.com/aws/aws-sdk-go/aws"
+	//  "github.com/aws/aws-sdk-go/aws/awserr"
+	"encoding/json"
+	"os"
+	"path"
+	"path/filepath"
+	"strconv"
+	"strings"
+	"sync"
+
+	"github.com/aws/aws-sdk-go/aws/awsutil"
+	"github.com/aws/aws-sdk-go/aws/credentials"
+	"github.com/aws/aws-sdk-go/aws/session"
+	"github.com/aws/aws-sdk-go/service/s3"
+)
+
+// definition of a worker instance
+type Worker struct {
+	Acl          string          // s3 acl for uploaded files - for our use either "public" or "private"
+	Bucket       string          // s3 bucket to upload to
+	Subfolder    string          // s3 subfolder destination (if needed)
+	Svc          *s3.S3          // instance of s3 svc
+	File_channel chan string     // the channel to get file names from (upload todo list)
+	Wg           *sync.WaitGroup // wait group - to signal when worker is finished
+	SourceDir    string          // where source files are to be uploaded
+	DestDir      string          // where to move uploaded files to (on local box)
+	Id           int             // worker id number for debugging
+}
+
+type PkgInfo struct {
+	Origin  string `json:"origin"`
+	Name    string `json:"name"`
+	Version string `json:"version"`
+	Release string `json:"release"`
+}
+
+// worker to get all files inside a directory (recursively)
+func get_file_list(searchDir string, file_channel chan string, num_workers int, wg *sync.WaitGroup) {
+	defer wg.Done() // signal we are finished at end of function or return
+
+	// sub function of how to recurse/walk the directory structure of searchDir
+	_ = filepath.Walk(searchDir, func(path string, f os.FileInfo, err error) error {
+
+		// check if it's a file/directory (we just want files)
+		file, err := os.Open(path)
+		if err != nil {
+			return nil
+		}
+
+		defer file.Close() // close file handle on return
+
+		fi, err := file.Stat()
+		if fi.IsDir() {
+			return nil
+		}
+
+		path = strings.Replace(path, searchDir, "", 1)
+		file_channel <- path // add file to the work channel (queue)
+		return nil
+	})
+
+	// add num_workers empty files on as termination signal to them
+	for i := 0; i < num_workers; i++ {
+		file_channel <- ""
+	}
+}
+
+// upload function for workers
+// uploads a given file to s3
+func (worker *Worker) upload(file string) (string, error) {
+	res, err := exec.Command("hab", "pkg", "info", "-j", fmt.Sprintf("%s%s", worker.SourceDir, file)).Output()
+
+	if err != nil {
+		panic(err)
```

Comment:
Emit a warning and return instead of panic? (assuming the panic exits the process)